sql/pipelines/src/main/scala/org/apache/spark/sql/pipelines/graph/DatasetManager.scala

```diff
@@ -0,0 +1,290 @@
+        .getDataflowGraph
+    } catch {
+      case e: SparkException if e.getCause != null => throw e.getCause
+    }
+
+    materializedGraph
+  }
+
+  /**
+   * Materializes a table in the catalog. This method will create or update the table in the
+   * catalog based on the given table and context.
+   * @param resolvedDataflowGraph The resolved [[DataflowGraph]] used to infer the table schema.
+   * @param table The table to be materialized.
+   * @param isFullRefresh Whether this table should be full refreshed or not.
+   * @param context The context for the pipeline update.
+   * @return The materialized table (with additional metadata set).
+   */
+  private def materializeTable(
+      resolvedDataflowGraph: DataflowGraph,
+      table: Table,
+      isFullRefresh: Boolean,
+      context: PipelineUpdateContext
+  ): Table = {
+    logInfo(log"Materializing metadata for table ${MDC(LogKeys.TABLE_NAME, table.identifier)}.")
+    val catalogManager = context.spark.sessionState.catalogManager
+    val catalog = (table.identifier.catalog match {
+      case Some(catalogName) =>
+        catalogManager.catalog(catalogName)
+      case None =>
+        catalogManager.currentCatalog
+    }).asInstanceOf[TableCatalog]
+
+    val identifier =
+      Identifier.of(Array(table.identifier.database.get), table.identifier.identifier)
+    val outputSchema = table.specifiedSchema.getOrElse(
+      resolvedDataflowGraph.inferredSchema(table.identifier).asNullable
+    )
+    val mergedProperties = resolveTableProperties(table, identifier)
+    val partitioning = table.partitionCols.toSeq.flatten.map(Expressions.identity)
+
+    val exists = catalog.tableExists(identifier)
+
+    // Error if partitioning doesn't match
+    if (exists) {
+      val existingPartitioning = catalog.loadTable(identifier).partitioning().toSeq
+      if (existingPartitioning != partitioning) {
+        throw new AnalysisException(
+          errorClass = "CANNOT_UPDATE_PARTITION_COLUMNS",
+          messageParameters = Map(
+            "existingPartitionColumns" -> existingPartitioning.mkString(", "),
+            "requestedPartitionColumns" -> partitioning.mkString(", ")
+          )
+        )
+      }
+    }
+
+    // Wipe the data if we need to
+    if ((isFullRefresh || !table.isStreamingTableOpt.get) && exists) {
+      context.spark.sql(s"TRUNCATE TABLE ${table.identifier.quotedString}")
+    }
+
+    // Alter the table if we need to
+    if (exists) {
+      val existingSchema = catalog.loadTable(identifier).schema()
+
+      val targetSchema = if (table.isStreamingTableOpt.get && !isFullRefresh) {
+        SchemaMergingUtils.mergeSchemas(existingSchema, outputSchema)
+      } else {
+        outputSchema
+      }
+
+      val columnChanges = diffSchemas(existingSchema, targetSchema)
+      val setProperties = mergedProperties.map { case (k, v) => TableChange.setProperty(k, v) }
+      catalog.alterTable(identifier, (columnChanges ++ setProperties).toArray: _*)
```

Comment:
Do we expect the catalog implementation to validate schema compatibility?

Reply:
`SchemaMergingUtils.mergeSchemas` above should validate this. There are some tests in `MaterializeTablesSuite` that exercise this.